### ISSUE TITLE
MethodInfo is null when evaluating a sample WASM

### ISSUE THREAD

**Issue by ebekker** [RyanLamansky/dotnet-webassembly#7]
Using [WebAssembly Studio](https://webassembly.studio/), I compiled the following sample C code to a WASM file:

`main.c:`
```c
#define WASM_EXPORT __attribute__((visibility("default")))

/* External function that is implemented in JavaScript. */
extern void sayc(char c);

int say_something(char *msg) {
  char *m = msg;
  while (*m)
    sayc(*(m++));

  return m - msg;
}

WASM_EXPORT
int main(void) {
  say_something("Hello World (from WASM)\n");
}
```

In this case I eliminated all references to any std C lib calls and only rely on a single externally defined function for generating output.  For completeness, here is a sample of the corresponding JavaScript file that defines this external func along with supporting code for invoking the WASM:

`main.js:`
```javascript
let x = '../out/main.wasm';

let instance = null;
let memoryStates = new WeakMap();

let s = "";
fetch(x).then(response =>
  response.arrayBuffer()
).then(bytes =>
  WebAssembly.instantiate(bytes, {
    env: {
      sayc: function (c) {
        c = String.fromCharCode(c);
        if (c == "\n") {
          console.log(s);
          s = "";
        } else {
          s += c;
        }
      }
    }
  })
).then(results => {
  instance = results.instance;
  document.getElementById("container").innerText = instance.exports.main();
}).catch(console.error);
```

Building this sample in Studio produces the WASM output that I'm attaching (as well as the S-Expression form).  (Attached as a ZIP due to GitHub restrictions.)
[main3.zip](https://github.com/RyanLamansky/dotnet-webassembly/files/2259780/main3.zip)

When running this in Studio, works as expected.  When running this using DNWASM, I'm getting this exception:

```
Unhandled Exception: WebAssembly.ModuleLoadException: At offset 56: Value cannot be null.
Parameter name: meth ---> System.ArgumentNullException: Value cannot be null.
Parameter name: meth
   at System.Reflection.Emit.ILGenerator.Emit(OpCode opcode, MethodInfo meth)
   at WebAssembly.CompilationContext..ctor(TypeBuilder exportsBuilder, FieldBuilder memory, Signature[] functionSignatures, MethodInfo[] methods, Signature[] types, Indirect[] functionElements, ModuleBuilder module, GlobalInfo[] globalGetters, GlobalInfo[] globalSetters) in C:\local\prj\bek\dotnet-webassembly\WebAssembly\CompilationContext.cs:line 96
   at WebAssembly.Compile.FromBinary(Reader reader, Type instanceContainer, Type exportContainer, IEnumerable`1 imports) in C:\local\prj\bek\dotnet-webassembly\WebAssembly\Compile.cs:line 412
   at WebAssembly.Compile.FromBinary[TExports](Stream input, IEnumerable`1 imports) in C:\local\prj\bek\dotnet-webassembly\WebAssembly\Compile.cs:line 61
   --- End of inner exception stack trace ---
   at WebAssembly.Compile.FromBinary[TExports](Stream input, IEnumerable`1 imports) in C:\local\prj\bek\dotnet-webassembly\WebAssembly\Compile.cs:line 85
   at Bek.Program.Main(String[] args) in C:\local\prj\bek\dotnet-webassembly\Bek\Program.cs:line 48
```

When I step through this, I see that MethodInfo argument is unresolved.

**Comment by RyanLamansky** [comment]
Fixing the precise issue you referenced is easy enough (the `functionElements` parameter should be passed as null at that stage), but your sample then runs into another situation where the compiler's internal tracking of function signatures has multiple flaws that become apparent in non-trivial situations.  The tests I had written were, unfortunately, "trivial", and didn't trigger the problem.

Putting in some hacks confirms that the signature tracking is the last issue; the code builds and runs perfectly if it gets through that portion.

It will take some effort to figure out the changes to signature tracking, maybe a few days depending on my free time.

**Comment by ebekker** [comment]
Thanks -- I realized when I re-read this issue, I had forgotten to include my C# side of the code, which is trivial enough and clearly you were able to craft your own sample, but I'll include it here for completeness and as a reference to others who may be interested:

`Program.cs:`
```csharp
    public class Program
    {
        public static void sayc(int c)
        {
            Console.Write(c);
            Console.Out.Flush();
        }

        public static MethodInfo GetMethod(string name)
        {
            var mi = typeof(Program).GetMethod(name);
            if (mi == null)
                throw new Exception("missing method " + name);
            return mi;
        }


        static void Main(string[] args)
        {
            var fileName = "main.wasm";

            Console.WriteLine("Hello World! (From .NET)");
            Console.WriteLine(File.Exists(fileName));

            using (var fs = File.OpenRead(fileName))
            {
                var instanceFunc = Compile.FromBinary<Program>(fs, imports: new[]
                {
                    new FunctionImport("env", "sayc",    GetMethod("sayc")),
                });
                var instance = instanceFunc();
            }
        }
    }
```

**Comment by RyanLamansky** [comment]
This should be fixed with 0.3.4-preview.  It's very cool to see C code running through it.

To defend against regressions, [I've added this code to the test suite](https://github.com/RyanLamansky/dotnet-webassembly/blob/1039580767dd0f7bbaceaf67ace0c04858dd441f/WebAssembly.Tests/SampleTests.cs).  A `StringBuilder` was used to collect the returned integers and recompose the original "Hello World (from WASM)\n".
